### PR TITLE
Feature/basic auth

### PR DESCRIPTION
**Pull request by sroy9675**
## Pull request type

-  Feature

## Description

Add Basic Authorization to the supported API Endpoints. Authorization will be required if the env variable "DATAFOG_AUTH_ENABLED" is set to true.  The authorized user and password must be set in an env variable or in a .env file under the keys: "DATAFOG_AUTH_USER" and "DATAFOG_PASSWORD" respectively

## Changes

- Add optional basic authorization to each endpoint.
- if the environment variable DATAFOG_AUTH_ENABLED is true basic authorization is required.  
- Valid username and password should be completely configured in the environment variables DATAFOG_AUTH_USER and DATAFOG_PASSWORD, or the same keys in a .env file, respectively (no mixing and matching allowed)
- created new authentication module to house authentication related code
- updated constants.py to include an ExceptionMessages Enum to manage the text displayed by exceptions
Issue Number: N/A

## Tests

created test_authentication.py to test new authentication module.
updated existing exception testing to use new ExceptionMessages Enum

## Other information

N/A

### CODE REVIEW

File: app/authorization.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+def get_authorization(credentials: Optional[HTTPBasicCredentials] = Depends(security)):
+    """Helper function to validate user authorization"""
+    if credentials is None:
+        return AuthTypes.NO_AUTH.value
+
+    if not is_valid_request(credentials):
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail=ExceptionMessages.UNAUTHORIZED.value,
+            headers={"WWW-Authenticate": "Basic"},
```
Comment (pselvana) [inline app/authorization.py]: "Basic" should be the mode/scheme that is being used instead of hardcoded.

File: app/authorization.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        )
+
+    return AuthTypes.HTTP_BASIC.value
```
Comment (pselvana) [inline app/authorization.py]: This should be the type returned as well.

File: app/authorization.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def is_valid_request(credentials: HTTPBasicCredentials) -> bool:
```
Comment (pselvana) [inline app/authorization.py]: rename to is_valid_basic_request

File: app/authorization.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    current_password_bytes = credentials.password.encode("utf8")
+    is_correct_username = secrets.compare_digest(
+        current_username_bytes, authorized_user_bytes
+    )
+    is_correct_password = secrets.compare_digest(current_password_bytes, password_bytes)
+
+    return is_correct_username and is_correct_password
+
+
+def load_valid_credentials() -> tuple[bytes, bytes]:
```
Comment (pselvana) [inline app/authorization.py:72]: Let's use:
from dotenv import load_dotenv
import os

load_dotenv()

PASSWORD_KEY = os.getenv('PASSWORD_KEY')

File: app/authorization.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    """read authentication configuration from env or fallback to file, throw if unsuccessful"""
+    # Note: configuration must be complete in either the env or the file, no mix and match
+    try:
+        authorized_user_bytes = os.environ[USER_KEY].encode("utf8")
+        password_bytes = os.environ[PASSWORD_KEY].encode("utf8")
+    except KeyError:
+        # auth config not set in env variables, attempt to read from .env file
```
Comment (pselvana) [inline app/authorization.py]: The dotenv solves this as well.

File: app/authorization.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    try:
+        authorized_user_bytes = config_dict[USER_KEY].encode("utf8")
+        password_bytes = config_dict[PASSWORD_KEY].encode("utf8")
+    except KeyError as exc:
+        # failed to find in file, throwing
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail=ExceptionMessages.AUTH_USER_KEY.value,
+            headers={"WWW-Authenticate": "Basic"},
```
Comment (pselvana) [inline app/authorization.py]: same as above on "Basic"

File: app/main.py [hunk 7]
```diff
@@ -1,9 +1,12 @@
 """API REST endpoints"""
 
+from typing import Optional
+
+from authorization import AUTH_ENABLED, get_authorization
```
Comment (pselvana) [inline app/main.py]: is there a pattern for seeing which files are local versus external?

Comment (pselvana) [reply app/main.py]: datafog-api. ?

File: app/main.py [hunk 8]
```diff
@@ -33,8 +39,11 @@ def annotate(
 def anonymize(
     text: str = Body(embed=True, min_length=1, max_length=1000, pattern=VALID_INPUT_PATTERN),
     lang: str = Body(embed=True, default="EN"),
+    user: Optional[str] = Depends(get_authorization),
 ):
     """entry point for anonymize functionality"""
+    if AUTH_ENABLED:
+        print(f"Authentication type: {user}")
```
Comment (pselvana) [inline app/main.py]: Is this user or is this auth_type?

File: app/main.py [hunk 9]
```diff
@@ -47,8 +56,11 @@ def encode(
     text: str = Body(embed=True, min_length=1, max_length=1000, pattern=VALID_INPUT_PATTERN),
     lang: str = Body(embed=True, default="EN"),
     salt: str = Body(embed=True, min_length=16, max_length=64),
+    user: Optional[str] = Depends(get_authorization),
 ):
     """entry point for reversible anonymize functionality"""
+    if AUTH_ENABLED:
+        print(f"Verified user: {user}")
```
Comment (pselvana) [inline app/main.py]: same as above.

**Review by pselvana — CHANGES_REQUESTED** [review]
Left comments to clean up basic to generalize and use dotenv.

File: app/authorization.py (hunk 2 above)
Comment (sroy9675) [reply app/authorization.py]: return the current active auth mode

File: app/main.py (hunk 9 above)
Comment (sroy9675) [reply app/main.py]: updated

File: app/main.py (hunk 8 above)
Comment (sroy9675) [reply app/main.py]: updated

File: app/authorization.py (hunk 4 above)
Comment (sroy9675) [reply app/authorization.py:72]: updated

File: app/authorization.py (hunk 5 above)
Comment (sroy9675) [reply app/authorization.py]: updated

File: app/authorization.py (hunk 6 above)
Comment (sroy9675) [reply app/authorization.py]: updated

File: app/authorization.py (hunk 3 above)
Comment (sroy9675) [reply app/authorization.py]: updated

File: app/authorization.py (hunk 1 above)
Comment (sroy9675) [reply app/authorization.py]: updated

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -3,7 +3,7 @@ repos:
     rev: 5.12.0
     hooks:
       - id: isort
-        args: [--profile=black]
+        args: [--profile=black, --src=app]
```
Comment (sroy9675) [inline .pre-commit-config.yaml:6]: help isort find all our local modules to correctly sort imports

Comment (pselvana) [reply .pre-commit-config.yaml:6]: Good find.

**Review by pselvana — APPROVED** [review]
LGTM! Nice cleanup.

**Pull request merged**